patterns/gitleaks/8.18.2/98-general.toml

```diff
@@ -528,9 +527,25 @@
       '''#\s*?noqa(:[\s\w]+)?$''',
     ]
 
+[[rules]]
+  id = 'e28317dbb1b2104c7fca5911047d8315152fd6f230bd07d5188fc9e461154559'
+  description = 'PyPI Upload Token'
+  regex = '''pypi-AgEIcHlwaS5vcmc[A-Za-z0-9-_]{50,1000}'''
+  tags = ['type:secret', 'alert:repo-owner']
+  keywords = [
+    'pypi-ageichlwas5vcmc',
+  ]
+
+[[rules]]
+  id = '76b4d3b89d4af399881204c7d8520e0c88a78f2662f31a3f674c51d4d64c8baa'
+  description = 'SendGrid API Key'
+  regex = '''\bSG\.[\w\-]{16,32}\.[\w\-]{16,64}\b'''
```

Comment:
`[\w\-]{16,64}\b'` wouldn't match `SG.abcd.abc-` — assuming it's possible for tokens to have a `-` at the end.

Reply:
Ack. Let me do some digging to see what's valid for a SG key. Thanks!

Reply:
I pulled some example keys from GitGuardian and GitHub (or really I found a lot of real SG keys and then replaced the data in them but keeping the structure :stuck_out_tongue: ) and It looks like this *should* catch them all. Also I peaked over at Trufflehog's regex and it looks roughly the same. 

Have you seen any keys in the wild this wouldn't catch?

Reply:
> Also I peaked over at Trufflehog's regex and it looks roughly the same.

I figured it was worth pointing out, as it can make a huge difference for some patterns (e.g., GitLab tokens). This is a problem with a number of patterns in TruffleHog's codebase — and far beyond.

> Have you seen any keys in the wild this wouldn't catch?

Not that I can recall!

Reply:
Thanks for the extra set of eyes ^_^ I'll roll it out this way and then probably end up doing additional tuning if we start getting a lot of FPs. The way the scanning service runs makes it easy to get feedback about the pattern quality :ok_hand: